cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java

```diff
@@ -1416,6 +1416,120 @@ private int moveLeadershipInBatch() throws InterruptedException, ExecutionExcept
+            LOG.debug("Task {} is marked as dead to force-stop the execution with a controller bounce.", task);
+            finishedTasks.add(task);
+            stoppedTaskIds.add(task.executionId());
+            _executionTaskManager.markTaskDead(task);
+          } else if (cluster.partition(tp) == null) {
+            handleProgressWithTopicDeletion(task, finishedTasks, deletedTaskIds);
+          } else if (ExecutionUtils.isLeadershipMovementDone(cluster, task)) {
+            handleProgressWithCompletion(task, finishedTasks);
+          } else {
+            if (shouldReportSlowTasks) {
+              task.maybeReportExecutionTooSlow(_time.milliseconds(), slowTasksToReport);
+            }
+            if (maybeMarkTaskAsDead(cluster, null, task, null)) {
+              deadTaskIds.add(task.executionId());
+              finishedTasks.add(task);
+            }
+          }
+        }
+        sendSlowExecutionAlert(slowTasksToReport);
+        updateOngoingExecutionState();
+      } while (!inExecutionTasks().isEmpty() && finishedTasks.isEmpty());
+
+      LOG.info("Finished tasks: {}.{}{}{}", finishedTasks,
+               stoppedTaskIds.isEmpty() ? "" : String.format(". [Force-Stopped: %s]", stoppedTaskIds),
+               deletedTaskIds.isEmpty() ? "" : String.format(". [Deleted: %s]", deletedTaskIds),
+               deadTaskIds.isEmpty() ? "" : String.format(". [Dead: %s]", deadTaskIds));
+    }
+
+    /**
+     * Periodically checks the metadata to see if intra-broker replica reassignment has finished or not.
+     */
+    private void waitForIntraBrokerReplicaTasksToFinish() {
+      List<ExecutionTask> finishedTasks = new ArrayList<>();
+      Set<Long> stoppedTaskIds = new HashSet<>();
+      Set<Long> deletedTaskIds = new HashSet<>();
+      Set<Long> deadTaskIds = new HashSet<>();
+      do {
+        // If there is no finished tasks, we need to check if anything is blocked.
+        maybeReexecuteIntraBrokerReplicaTasks();
+        Cluster cluster = getClusterForExecutionProgressCheck();
+        Map<ExecutionTask, ReplicaLogDirInfo> logDirInfoByTask = getLogdirInfoForExecutionTask(
+            _executionTaskManager.inExecutionTasks(Collections.singleton(INTRA_BROKER_REPLICA_ACTION)),
+            _adminClient, _config);
+
+        List<ExecutionTask> slowTasksToReport = new ArrayList<>();
+        boolean shouldReportSlowTasks = _time.milliseconds() - _lastSlowTaskReportingTimeMs > SLOW_TASK_ALERTING_BACKOFF_TIME_MS;
+        for (ExecutionTask task : inExecutionTasks()) {
+          TopicPartition tp = task.proposal().topicPartition();
+          if (_stopSignal.get() == FORCE_STOP_EXECUTION) {
+            // If the execution is force-stopped, the executor will mark all in progress tasks as dead, cleanup the
+            // relevant reassignment zNodes, and then bounce the controller (see ExecutionUtils#deleteZNodesToStopExecution).
```

Comment:
Is this comment correct? I thought CC no longer deletes the znode and moves controller in order to stop the execution.

Reply:
Please see the response to the other comment above.